cranelift-codegen/src/isa/x86/enc_tables.rs

```diff
@@ -895,6 +895,84 @@ fn expand_fcvt_to_uint_sat(
     cfg.recompute_ebb(pos.func, done);
 }
 
+/// Expand shuffle instructions
+fn expand_shuffle(
+    inst: ir::Inst,
+    func: &mut ir::Function,
+    _cfg: &mut ControlFlowGraph,
+    _isa: &dyn TargetIsa,
+) {
+    let mut pos = FuncCursor::new(func).at_inst(inst);
+    pos.use_srcloc(inst);
+
+    if let ir::InstructionData::Shuffle { args, mask, .. } = pos.func.dfg[inst] {
+        // remove the original shuffle instruction
+        pos.remove_inst();
+
+        // extract the original result value and use it (e.g. with_result) as the final result value of our replaced sequences
+        let original_results = pos.func.dfg.detach_results(inst);
+        let original_result = original_results.get(0, &pos.func.dfg.value_lists).unwrap();
+
+        // a mask-building helper: in 128-bit SIMD, 0-15 indicate which lane to read from and a 1 in the most significant position zeroes the lane
+        let build_shuffle_mask = |b: u8| if b > 15 { 0b10000000 } else { b };
+
+        // we only have to worry about aliasing here because copies will be introduced later (in regalloc)
+        let a = pos.func.dfg.resolve_aliases(args[0]);
+        let b = pos.func.dfg.resolve_aliases(args[1]);
+        let mask = pos
+            .func
+            .dfg
+            .immediates
+            .get(mask)
+            .expect("The shuffle immediate should have been recorded before this point")
+            .clone();
+        if a == b {
+            // PSHUFB the first argument (since it is the same as the second)
+            let constructed_mask = mask
+                .iter()
+                .cloned()
+                .map(|b| if b > 15 { b.wrapping_sub(16) } else { b }) // if the mask is greater than 15 it still may be referring to a lane in b
+                .map(build_shuffle_mask)
+                .collect();
+            let handle = pos.func.dfg.constants.insert(constructed_mask);
+            let a_type = pos.func.dfg.value_type(a);
+            let mask_value = pos.ins().vconst(a_type, handle); // move the built mask into another XMM register
+            pos.ins()
+                .with_result(original_result)
+                .x86_pshufb(args[0], mask_value); // shuffle the single incoming argument
```

Comment:
I'm not a big fan of early return unless it's at the very top of the function or everything is very simple. I'm concerned that it will be hard to see in this rather large function and confuse someone later.